websearch/brave/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+mod client;
+mod conversions;
+
+use std::cell::RefCell;
+
+use crate::client::{BraveSearchApi, SearchRequest};
+use crate::conversions::{params_to_request, response_to_results, validate_search_params};
+use golem_web_search::golem::web_search::web_search::{
+    Guest, GuestSearchSession, SearchError, SearchMetadata, SearchParams, SearchResult,
+    SearchSession,
+};
+
+use golem_web_search::LOGGING_STATE;
+
+struct BraveSearch {
+    client: BraveSearchApi,
+    request: SearchRequest,
+    params: SearchParams,
+    finished: bool,
+    metadata: Option<SearchMetadata>,
+    current_offset: u32,
+}
+
+impl BraveSearch {
+    fn new(client: BraveSearchApi, request: SearchRequest, params: SearchParams) -> Self {
+        Self {
+            client,
+            request,
+            params,
+            finished: false,
+            metadata: None,
+            current_offset: 0,
+        }
+    }
+
+    fn next_page(&mut self) -> Result<Vec<SearchResult>, SearchError> {
+        if self.finished {
+            return Ok(vec![]);
+        }
+
+        // Update request with current offset
+        let mut request = self.request.clone();
+        request.offset = Some(self.current_offset);
+
+        let response = self.client.search(request)?;
+        let (results, metadata) = response_to_results(response, &self.params, self.current_offset);
+
+        // Check if more results are available
+        if let Some(ref meta) = metadata {
+            // Check if we got the full count requested - if yes, there might be more
+            let count = self.request.count.unwrap_or(10);
+            let has_more_results = results.len() == (count as usize);
+
+            // Also check if next_page_token is available
+            let has_next_page = meta.next_page_token.is_some();
+
+            // Only set finished if no more results available
+            self.finished = !has_more_results || !has_next_page;
+
+            // Increment offset for next page if not finished
+            if !self.finished {
+                self.current_offset += count;
```

Comment:
If I'm reading the docs correctly this should be incremented by 1, not count:

https://api-dashboard.search.brave.com/app/documentation/web-search/query